### PR TITLE
purescript-bridge compatibility changes 

### PR DESCRIPTION
**Pull request by gnumonik**
Hi CTL folks, 

This PR contains fixes/changes to the `ToData/FromData` machinery that are necessary for the correct operation of `purescript-bridge`. The changed modules should be well-documented, but the general idea is: 

When one extracts a `RowList` from a `Row`, the `RowList` is always lexicographically sorted. Plutus Data records are _not_ lexicographically sorted, so the existing generic To/FromData code in CTL does not work correctly for Records right now. Because both Haskell and PureScript permit (for some reason...seriously who does this?!) data types such as this gem

``` haskell
data Foo = FooA {foo :: String, fooo :: Int}
                | FooB {fooo :: Bool, foo :: Bool}
```
We have to capture the relation between a constructor and its record fields (if any exist) in order to ensure on-chain compatibility. In order to integrate with most of the existing generic To/FromData code, this has to be done at the type level, so that's where we do it. 

I will attempt to pull/merge CTL's `master` branch ASAP to resolve conflicts, but that may take me a day (this is by far the most complex merge I've attempted, I will probably screw it up at least once). In the meantime, please review this PR and suggest relevant/necessary changes. 

Edit: 

Actually there are way fewer conflicts than I expected. Will come back to this first thing tomorrow and sort that out.

### CODE REVIEW

**Review by bladyjoker — CHANGES_REQUESTED** [review]
@gnumonik let's add the ToData/FromData instances for the Interval type as we need it and it's a good place as any to demo the PlutusSchema

File: src/FromData.purs [hunk 1]
```diff
@@ -2,190 +2,253 @@ module FromData
...
 import Data.Ratio (Ratio, reduce)
+import Data.Show.Generic (genericShow)
 import Data.Symbol (class IsSymbol, reflectSymbol)
 import Data.TextDecoder (decodeUtf8)
 import Data.Traversable (traverse)
 import Data.Tuple (Tuple(Tuple))
 import Data.UInt (UInt)
 import Data.Unfoldable (class Unfoldable)
 import Helpers (bigIntToUInt)
+import Prelude
```
Comment (ngua) [inline src/FromData.purs:35]: Normally I'd encourage explicit imports, but for `Prelude` I don't think it's necessary (and makes it harder to keep track of redundant or missing imports)

File: src/TypeLevel/DataSchema.purs [hunk 2]
```diff
@@ -0,0 +1,238 @@
+module TypeLevel.DataSchema
```
Comment (ngua) [inline src/TypeLevel/DataSchema.purs:1]: I'd personally change the namespace of the module from `TypeLevel` to `Plutus`, or even `Plutus.Types` (so possible `Plutus.Types.DataSchema`), just to make it clear that this is quite Plutus-specific

File: src/TypeLevel/DataSchema.purs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+  , I
+  , type (:+)
+  , IxK
+  , MkIxK
+  , MkIxK_
+  , type (@@)
+  , Field
+  , MkField
+  , MkField_
+  -- , NoRec
```
Comment (ngua) [inline src/TypeLevel/DataSchema.purs:17]: Why is this commented out?

File: src/TypeLevel/RowList.purs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  ) =>
+  AllUniqueLabels (Cons k a (Cons k' a' xs))
+
+-- | Poor man's type level tests
+tests ∷ Array String
+tests =
+  [ testNil
+  , testSingleton
+  , testUniques
+  -- , testDupsUnordered
```
Comment (ngua) [inline src/TypeLevel/RowList.purs:30]: Why are these commented out? I see the definitions aren't

File: src/TypeLevel/RowList.purs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  Fail (Text "Labels are not unique!") =>
+  AllUniqueLabels (Cons k a (Cons k a' xs))
+else instance
+  ( AllUniqueLabels (Cons k a xs)
+  , AllUniqueLabels (Cons k' a' xs)
+  ) =>
+  AllUniqueLabels (Cons k a (Cons k' a' xs))
+
+-- | Poor man's type level tests
+tests ∷ Array String
```
Comment (ngua) [inline src/TypeLevel/RowList.purs:25]: Could these be moved to `test/Data.purs`?

File: src/TypeLevel/RowList/Unordered/Indexed.purs [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+  Fail (Text "Indices are not unique!") =>
+  UniqueIndices (ConsI k a n (ConsI k' a' n xs))
+else instance
+  ( UniqueIndices (ConsI k a n xs)
+  , UniqueIndices
+      (ConsI k' a' n' xs)
+  ) =>
+  UniqueIndices (ConsI k a n (ConsI k' a' n' xs))
+
+uniqueIndicesTests ∷ Array String
```
Comment (ngua) [inline src/TypeLevel/RowList/Unordered/Indexed.purs:52]: Same here

File: test/Data.purs [hunk 7]
```diff
@@ -2,15 +2,24 @@
 module Test.Data (suite) where
 
 import Prelude
```
Comment (ngua) [inline test/Data.purs:4]: See my other comment about `Prelude` imports

**Review by ngua — COMMENTED** [review]
First, this is exceptionally well documented, thank you for taking the time to write out the rationale behind this approach. It definitely helped me understand why such large changes were necessary

My biggest concern is the potential cognitive load on library users. Could we have more examples in the documentation of `HasPlutusSchema`? I know we have the `Foo` example but I think more would be good. It would also be nice if we could document migrating from the `HasConstrIndices` approach to this one

I have some other minor suggestions below as well

File: src/TypeLevel/RowList.purs (hunk 4 above)
Comment (gnumonik) [reply src/TypeLevel/RowList.purs:30]: @bladyjoker implemented these tests so he should probably say whether he thinks the commented test is still needed, but I'll move everything to a more appropriate module in the next set of commits

File: src/TypeLevel/DataSchema.purs (hunk 3 above)
Comment (gnumonik) [reply src/TypeLevel/DataSchema.purs:17]: Dead code (predates the fancy syntactic sugar in DataSchema). Will remove.

File: src/TypeLevel/RowList.purs (hunk 4 above)
Comment (gnumonik) [reply src/TypeLevel/RowList.purs:30]: Oh now I remember: Drazen missed an instance when reviewing and thought the compile-time checks worked a bit differently than they actually do. Nothing with duplicates should/does compile. I'm going to remove those.

**Comment by gnumonik** [comment]
I made most of the requested changes, still need to throw together a diverse set of examples for the schema stuff, will do that over the weekend.

File: src/TypeLevel/RowList.purs (hunk 4 above)
Comment (bladyjoker) [reply src/TypeLevel/RowList.purs]: The problem with type level programming, is that you can't test your type level programs :)

File: src/ToData.purs [hunk 8]
```diff
@@ -44,110 +51,179 @@ class ToData :: Type -> Constraint
 class ToData a where
   toData :: a -> PlutusData
 
-class ToDataWithIndex :: Type -> Type -> Constraint
-class HasConstrIndices ci <= ToDataWithIndex a ci where
-  toDataWithIndex :: Proxy ci -> a -> PlutusData
+{- | A class which converts a type to its PlutusData representation with the help of a
```
Comment (klntsky) [inline src/ToData.purs:54]: You need to prefix each line of doc-comment with `|`:

https://github.com/purescript/documentation/blob/master/language/Syntax.md#comments

> Note that, unlike Haskell, every line which should be considered documentation must start with a pipe. This allows you to do things like:

**Review by klntsky — CHANGES_REQUESTED** [review]

File: test/Data.purs [hunk 9]
```diff
@@ -341,3 +455,56 @@ testBinaryFixture value binaryFixture = do
     map (toBytes <<< asOneOf) (PDS.convertPlutusData (toData value))
       `shouldEqual` Just
         (hexToByteArrayUnsafe binaryFixture)
+
+-- | Poor man's type level tests
+tests ∷ Array String
+tests =
```
[the remark this answers is not in the document]
Comment (bladyjoker) [reply test/Data.purs:461]: nit; If you can align the 2 `type tests` as they're essentially the same

File: src/TypeLevel/RowList/Unordered/Indexed.purs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+  ( AllUniqueLabelsI (ConsI k a n xs)
+  , AllUniqueLabelsI (ConsI k' a' n' xs)
+  ) =>
+  AllUniqueLabelsI (ConsI k a n (ConsI k' a' n' xs))
+
+-- | Get the Nat index corresponding to a given symbol index in a RowListI.
+-- This is meant to be used primarily in the context of an instance declaration for some
+-- other class  (see examples in ToData.purs), and is probably not particularly useful
+-- outside of those instance contexts.
+class GetIndexWithLabel
```
Comment (bladyjoker) [inline src/TypeLevel/RowList/Unordered/Indexed.purs:71]: Add type-level tests

File: src/TypeLevel/RowList/Unordered/Indexed.purs [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+  ( UniqueIndices (ConsI k a n xs)
+  , UniqueIndices
+      (ConsI k' a' n' xs)
+  ) =>
+  UniqueIndices (ConsI k a n (ConsI k' a' n' xs))
+
+-- | Uniqueness constraint on the labels of a RowListI which asserts that all labels are unique.
+-- Again, this is needed so that the lookup functions perform in the expected manner.
+class AllUniqueLabelsI :: forall (k :: Type). RowListI k -> Constraint
+class AllUniqueLabelsI list
```
Comment (bladyjoker) [inline src/TypeLevel/RowList/Unordered/Indexed.purs:54]: Add type-level tests

File: src/TypeLevel/RowList/Unordered/Indexed.purs [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+  , IsSymbol label
+  , KnownNat ix
+  ) =>
+  GetLabelWithIndex ix (ConsI label' a' n' xs) label
+
+-- | Given a Symbol which appears as a label in the given RowListI, get ahold of the type indexed by that symbol.
+-- Note that this does not "return" a *value* of that type, but the type itself, and is therefore only suitable
+-- for type class instance contexts (& other type level computational contexts if any exist)
+class GetWithLabel
+  :: forall (k :: Type). Symbol -> RowListI k -> k -> Constraint
```
Comment (bladyjoker) [inline src/TypeLevel/RowList/Unordered/Indexed.purs:123]: Add type-level tests

File: src/TypeLevel/RowList/Unordered/Indexed.purs [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+class GetWithLabel label rlist result | label rlist -> result
+
+instance IsSymbol l => GetWithLabel l (ConsI l a n xs) a
+else instance (GetWithLabel l xs a) => GetWithLabel l (ConsI _l _a _n xs) a
+
+-- | Get the nat index at a symbol in a RowList (Note: Not a RowListI! Use GetIndexWithLabel for a RowListI)
+class GetLabelIndex
+  :: forall (k :: Type). Symbol -> RowList k -> Nat -> Constraint
```
Comment (bladyjoker) [inline src/TypeLevel/RowList/Unordered/Indexed.purs:131]: Add type-level tests

**Comment by gnumonik** [comment]
I will make changes / transfer the PR to a branch on the main repo first thing tomorrow. Got caught up with some last minute stuff in `purescript-bridge`, out of time/energy for today.

File: src/TypeLevel/DataSchema.purs (hunk 2 above)
Comment (bladyjoker) [reply src/TypeLevel/DataSchema.purs]: Done

File: src/TypeLevel/RowList.purs (hunk 5 above)
Comment (bladyjoker) [reply src/TypeLevel/RowList.purs]: Done

File: src/ToData.purs (hunk 8 above)
Comment (bladyjoker) [reply src/ToData.purs:54]: Will do this in #417

File: src/TypeLevel/RowList/Unordered/Indexed.purs (hunk 10 above)
Comment (bladyjoker) [reply src/TypeLevel/RowList/Unordered/Indexed.purs:71]: Not deemed necessary

File: test/Data.purs (hunk 7 above)
Comment (bladyjoker) [reply test/Data.purs:4]: Done

**Comment by bladyjoker** [comment]
Closed in favor of #417

**Pull request closed without merging**